feat(tutor): set up ai tutor courses pilot experiment

drizco opened the pull request:
This pull request refactors and standardizes the logic for determining when the AI Tutor should be shown to users, introduces a new pilot flag for AI Tutor availability, and ensures this logic is consistently applied across the codebase.

Key changes include:

**AI Tutor Availability Logic Refactor:**

* Refactored the `shouldShowAiTutor` helper to accept an object with explicit parameters (`appName`, `tutorLevel`, and `tutorPilot`), improving clarity and maintainability. The helper was moved to `apps/src/aiTutor/helpers/shouldShowAiTutor.ts` and updated throughout the codebase to use the new signature.
* Added unit tests for `shouldShowAiTutor` to ensure correct behavior for all combinations of input flags.

**AI Tutor Pilot Experiment Support:**

* Added a new `aiTutorEnabledForPilot` flag to `appOptions` in the backend, with logic to determine pilot participation for teachers and students (students inherit from teachers). This is exposed to the frontend for use in AI Tutor visibility checks.
* Updated frontend utilities to read the new `aiTutorEnabledForPilot` flag from `appOptions` and provide a `getAiTutorEnabledForPilot` helper.
* Added tests to verify the backend logic for pilot flag assignment for both teachers and students.

**Integration and Consistency:**

* Updated all relevant React components and legacy lab views to use the new AI Tutor visibility logic, ensuring consistent AI Tutor display based on the pilot flag and level properties.

These changes ensure that AI Tutor is shown only to users who are in the pilot and on eligible levels, and that the logic is robustly tested and consistent across the application.

<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->



## Links
<!--  Links to relevant external resources.
      - spec docs, Jira tickets, related PRs, Honeybadger errors, etc.
-->

- Jira: https://codedotorg.atlassian.net/browse/LABS-1713

## Testing story
<!--  Does your change include appropriate tests?
      - If so, please describe how the tests included in this PR are sufficient.
      - If not, please explain why this change does not need to be tested. -->



## Deployment strategy
<!--  Any special steps required to deploy this, besides merging?
      - Are there database migrations or manual steps?
      - Does this require coordination with other teams or systems? -->



## Follow-up work
<!--  List any clean-up or technical debt that will be addressed in future work.
      - Include Jira links where applicable.
      - Are there any known limitations or improvements planned? -->



## Privacy
<!--  How does this change impact Student & Teacher privacy?
      - Does this collect, use, share, or modify PII, either new or existing? -->



## Security
<!--  How does this change impact our security surface-area?
      - Do API's touched have the right auth?
      - Do infra changes impact our attack surface or encryption?
-->



## Caching
<!--  How does this change impact caching behavior?
      - Are cache keys or invalidation strategies affected?
      - Will this require cache warming or invalidation after deployment? -->



## PR Creation Checklist:
<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy impacts have been documented
- [ ] Security impacts have been documented
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

Commit:
feat: add appOptions property for ai tutor access determined by teacher joining pilot or student of teacher that joined the pilot

Commit:
feat: update helper to check pilot and project level and add unit tests

Commit:
feat: use window.appOptions to check if ai tutor is available and enabled for the user in the pilot

Commit:
feat: check appOptions via script tag for lab2 labs

Commit:
refactor: move helper to aiTutor directory and move query param checking to resource panel and python lab so helper can be used in legacy labs

Commit:
feat: remove isProjectLevel client side check

relying entirely on level attribute ai_tutor_available. project levels will be updated via one off script

Commit:
chore: pr review updates

Commit:
feat: short circuit the experiments lookup if the level isn't ai tutor enabled anyway

cnbrenci commented on dashboard/app/helpers/levels_helper.rb:
```diff
@@ -545,6 +546,24 @@ def set_puzzle_position_options(level_options)
+  # Teachers: enabled if they are in the pilot single user experiment
+  # Students: enabled if any of their teachers are in the pilot
+  def ai_tutor_enabled_for_pilot?(user = current_user, level = @level)
+    return false unless user && level&.ai_tutor_available?
```
should this be here? `&& level&.ai_tutor_available?`

cnbrenci replied:
I think it's OK for the pilot check to be in appOptions like this for lab2 but if we are doing a level-specific check here that would cause a bug switching between lab2 levels?

cnbrenci replied:
Feels like this might just be a leftover artifact from a previous implementation that slipped through, because I see the rest of the pr treating the pilot vs level check separately

drizco replied:
I can remove it. I added this late to short circuit the experiment query if tutor isn't available on the level anyway.

drizco replied:
I'm curious why switching lab2 levels could cause a bug though? Is the `@level` variable not getting updated by the controller between lab2 levels?

cnbrenci replied:
since the page doesn't reload between lab2 levels, aiTutorEnabledForPilot in appOptions will be stale, based on the first level you loaded into. So if you have a progression with 2 adjacent lab2 levels, and the first doesn't have ai_tutor_available, you wouldn't see tutor on the second level.

drizco replied:
oh okay! this makes sense. I'm thinking it's okay to keep this in appOptions after removing the level short circuit, since checking if a student's teacher is in the pilot will not change, probably until we remove the pilot altogether. I didn't want to put it on the user or in the cookies since it would need to query the teachers' experiments each time.

thank you so much for catching this!

cnbrenci replied:
I agree! You're welcome :) definitely want to give credit to Sanchit and Emily who jumped on my request for a couple more student learning eyeballs and Sanchit pointed it out!

cnbrenci commented on apps/src/templates/instructions/InstructionsWithWorkspace.jsx:
```diff
@@ -113,9 +114,22 @@ export class UnwrappedInstructionsWithWorkspace extends React.Component {
+
+    // level property says ai tutor is available
+    const aiTutorAvailableForLevel =
+      window?.appOptions?.level?.aiTutorAvailable ?? false;
```
here should be fine since these are in legacy labs that do reload between levels, so appOptions would update

cnbrenci commented on apps/src/pythonlab/PythonlabView.tsx:
```diff
@@ -108,13 +112,15 @@ const PythonlabView: React.FunctionComponent<
+    shouldShowAiTutor({
+      tutorPilot: getAiTutorEnabledForPilot(),
+      appName: levelProperties.appName,
+      tutorLevel: levelProperties.aiTutorAvailable,
```
here you're using levelProperties, which I believe is updating the value on level change in `summarize_for_lab2_properties` so you should be good there - that's the supported way to access level properties in lab2

cnbrenci commented on apps/src/pythonlab/PythonlabView.tsx:
```diff
@@ -108,13 +112,15 @@ const PythonlabView: React.FunctionComponent<
+
+  const isAiTutorEnabled =
+    shouldShowAiTutor({
+      tutorPilot: getAiTutorEnabledForPilot(),
```
This will be stale between lab2 levels, BUT that's OK if you're not looking at level-specific data. This is where the bug would be if you're short-circuiting `ai_tutor_enabled_for_pilot` with the level property

Commit:
fix: remove short circuit level check for ai_tutor_available_for_pilot? method

lab2 levels do not re-fetch appOptions on level changes so ai tutor being available on the level would be stale

cnbrenci reviewed (approved):
🎉
